inference-engine/include/cpp/ie_cnn_network.h

```diff
@@ -54,8 +55,10 @@ class INFERENCE_ENGINE_API_CLASS(CNNNetwork) {
      * This constructor wraps existing ngraph::Function
      * If you want to avoid modification of original Function, please create a copy
      * @param network Pointer to the ngraph::Function object
+     * @param exts Vector of pointers to IE extension objects
      */
-    explicit CNNNetwork(const std::shared_ptr<ngraph::Function>& network);
+    explicit CNNNetwork(const std::shared_ptr<ngraph::Function>& network,
+                        const std::vector<IExtensionPtr>& exts = {});
```

Comment:
I think about the case if user creates nGraph Function from the code. Should he also create vector `IExtensionPtr` or map with opsets should be enough? Also the map of custom operation sets and custom opset aren't needed (I mean that from this perspective user should do something in anyway: creates vector of extensions or map of opsets) if user just creates nGraph Function, we need it only for reader and for serialization.

From other hand, in this case (when we pass vector of extensions) CNNNetwork holds all extension libraries and user cannot unload extension libraries.

@jdanieck  @ilya-lavrenov What do you think?

Reply:
Why do we need a vector of extensions if we create `ngraph::Function` from code? All custom operations should be already used in `ngraph::Function` and no needs to keep it inside `CNNNetwork`, is not it?

Reply:
Operation does not contain opset name which is needed to generate proper IR. Opset name is retrieved by searching for op in opsets objects, and custom opsets are stored in extension object.

Reply:
looks like a WA..

Reply:
Do you see better way to solve this? Proper solution would be to store opset name in ngraph op, but this would lead to major codebase refactoring (I'm thinking mostly about tests).

Reply:
Also we can add opset name to rt_info section. It should allow to avoid any additional arguments in the CNNNetwork and serializer. But in this case we need to be sure that out transformations don't corrupt this field in rt_info.

Reply:
Can you elaborate on rt_info ? This is node property right? So I could modify Readers to store this info there, but how this would work with programmatically created ngraph::Function?

Reply:
Yes rt_info is a node property. We use it in order to store some runtime information. And if user creates a node from the code we cannot add automatically opset name to runtime information.

But if user uses method `opset.create()` we can use this method to add automatically an opset name to rt_info. In this case we need to extend API of opset class, but also it allows to avoid any changes in readers.

Reply:
I think that currently passing handle to Extension objects in CNNNetwork constructor is the cleanest approach. If user wants to have custom operations he need to create Extension object either way and register it in IE. Programmatically creating ngraph::Function is at slightly lower abstraction level, but it should not be surprising that we need to pass this object also here. In case there are no custom ops we can omit this parameter since it's empty by default.

Reply:
So, @ilyachur you want me to explore rt_info approach? If so, then this PR will take longer and we won't be able to merge it before FF.

Reply:
I propose to don't think about the feature freeze, first of all we need to provide a good solution 😄 

From my point of view it is a hack, if we use rt_information. The original idea of rt_info structure is that it allows to store some runtime information.
I mentioned this approach only because it allow to do the same things without API changes. 
But if we use this approach we should be very careful, because some transformations can corrupt this information and in this case we cannot recover the original opset.

But from another hand if we pass Extension ptr to the CNNNetwork, looks like we pass a lot of redundant informations (for example extensions also can contains execution kernels for specific plugins).

If we pass only map of Operation sets, it means that we cannot control the life time of the runtime which has created this map and as a result can get a crash.

Reply:
How about some sort of a container for custom opsets in inference engine? This container could be filled with custom opsets whenever an extension is registered by the user and our API would be responsible for doing it. This way we would still be able to access those custom opsets and ops for any reason, iterate over them and perhaps use them to deserialize an XML containing custom ops.
Of course such custom opsets need to be removed from a container when an extension is unloaded.
I think this way we keep separation of concerns, don't have to modify the CNNNetwork class and the serialization mechanism is still able to access everything it needs to work.

Reply:
@tomdol We have an IE Core object which is a container for Inference Engine plugins, and we use it to register Inference Engine Extensions (and custom operations). This objects contains all registered extensions.

But at the current moment CNNNetwork is independent object and doesn't depend on IE Core.
And it means that when we call `serialize()` method, CNNNetwork doesn't have an information about custom operation sets and also cannot get this information from the Inference Engine Core object.

Reply:
Also as I can see for nGraph transformation we use the map of operation sets as an argument.

I thought about life time of Extension, and I think it is ok, if we don't hold extensions inside the CNNNetwork, because Core object already holds all these extensions.

Reply:
Ok, we discussed the issue. Let's keep the current implementation, because it allow also allow us to avoid some potential issues with life time of runtime libraries.